File: tests/samples/java/VulnerableApp.java
/**
 * Intentionally Vulnerable Java Application — Scanner Test Target
 * Contains deliberate vulnerabilities to validate scanner detection.
 * DO NOT DEPLOY. FOR TESTING ONLY.
 */

import java.io.*;
import java.sql.*;
import java.security.MessageDigest;
import javax.xml.parsers.*;
import java.nio.file.*;
import java.net.*;

public class VulnerableApp {

    // ── Hardcoded Secrets (JAVA-A02-001) ─────────────────────────────────
    private static final String DB_PASSWORD = "SuperSecret123!";   // VULN
    private static final String API_KEY     = "sk-prod-abc123";    // VULN
    private String token = "hardcoded-jwt-secret";                  // VULN

    // ── SQL Injection (JAVA-A03-001) ──────────────────────────────────────
    public ResultSet getUserById(Connection conn, String userId) throws SQLException {
        // VULN: string concatenation in executeQuery
        Statement stmt = conn.createStatement();
        return stmt.executeQuery("SELECT * FROM users WHERE id = " + userId);
    }

    public void updateEmail(Connection conn, String userId, String email)
            throws SQLException {
        // VULN: string concatenation in prepareStatement
        PreparedStatement ps = conn.prepareStatement(
            "UPDATE users SET email = '" + email + "' WHERE id = " + userId);
        ps.executeUpdate();
    }

    // ── Weak Crypto (JAVA-A02-002) ────────────────────────────────────────
    public String hashPasswordInsecure(String password) throws Exception {
        MessageDigest md = MessageDigest.getInstance("MD5");     // VULN: MD5
        byte[] hash = md.digest(password.getBytes());
        return bytesToHex(hash);
    }

    public String signatureInsecure(String data) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-1");   // VULN: SHA-1
        return bytesToHex(md.digest(data.getBytes()));
    }

    // ── XXE Injection (JAVA-A05-001) ─────────────────────────────────────
    public void parseXmlInsecure(InputStream xmlInput) throws Exception {
        // VULN: DocumentBuilderFactory without XXE mitigations
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        builder.parse(xmlInput);
    }

    public void parseSaxInsecure(InputStream xmlInput) throws Exception {
        // VULN: SAXParserFactory without XXE mitigations
        SAXParserFactory factory = SAXParserFactory.newInstance();
        SAXParser parser = factory.newSAXParser();
    }

    // ── Path Traversal (JAVA-A01-001) ─────────────────────────────────────
    public String readUserFile(String filename) throws IOException {
        // VULN: File constructed with user input via concatenation
        File file = new File("/app/uploads/" + filename);
        return new String(Files.readAllBytes(file.toPath()));
    }

    public byte[] downloadReport(String reportName) throws IOException {
        // VULN: Paths.get with concatenation — no canonicalization
        Path path = Paths.get("/app/reports/" + reportName);
        return Files.readAllBytes(path);
    }

    // ── Helper ─────────────────────────────────────────────────────────────
    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) sb.append(String.format("%02x", b));
        return sb.toString();
    }
}

File: tests/samples/python/vulnerable_app.py
"""
Intentionally Vulnerable Python Application — Scanner Test Target
=================================================================
This file contains deliberate vulnerabilities to validate scanner detection.
DO NOT deploy this code. FOR TESTING ONLY.
"""

import hashlib
import subprocess
import pickle
import sqlite3
import yaml
from flask import Flask, request, jsonify

app = Flask(__name__)

# ── Hardcoded Secrets (PY-A02-001) ──────────────────────────────────────────
DB_PASSWORD   = "SuperSecret123!"           # VULN: hardcoded password
API_KEY       = "sk-prod-abc123xyz789"      # VULN: hardcoded API key
SECRET_KEY    = "my-flask-secret"           # VULN: hardcoded flask secret
app.config["SECRET_KEY"] = SECRET_KEY

# ── Broken Crypto (PY-A02-002) ──────────────────────────────────────────────
def hash_password_insecure(password: str) -> str:
    return hashlib.md5(password.encode()).hexdigest()   # VULN: MD5

def verify_token_insecure(token: str) -> str:
    return hashlib.sha1(token.encode()).hexdigest()     # VULN: SHA-1


# ── SQL Injection (PY-A03-001) ───────────────────────────────────────────────
@app.route("/users")
def get_user():
    user_id = request.args.get("id")
    conn = sqlite3.connect("app.db")
    cur  = conn.cursor()

    # VULN: f-string SQL injection
    query = f"SELECT * FROM users WHERE id = {user_id}"
    cur.execute(query)

    # VULN: % format SQL injection
    query2 = "SELECT * FROM orders WHERE user_id = %s" % user_id
    cur.execute(query2)

    # VULN: .format() SQL injection
    query3 = "SELECT email FROM users WHERE name = '{}'".format(user_id)
    cur.execute(query3)

    return jsonify(cur.fetchall())


# ── Command Injection (PY-A03-002) ──────────────────────────────────────────
@app.route("/ping")
def ping_host():
    host = request.args.get("host")

    # VULN: shell=True with variable
    result = subprocess.run(f"ping -c 1 {host}", shell=True, capture_output=True)
    return result.stdout.decode()


# ── Insecure Deserialization (PY-A08-001) ────────────────────────────────────
@app.route("/load", methods=["POST"])
def load_data():
    data = request.get_data()

    # VULN: pickle.loads on untrusted data
    obj = pickle.loads(data)

    # VULN: yaml.load without SafeLoader
    config = yaml.load(data, Loader=yaml.FullLoader)

    return jsonify({"status": "loaded"})


# ── Missing Authentication (PY-A01-001) ──────────────────────────────────────
@app.route("/admin/users", methods=["GET"])        # VULN: no auth decorator
def admin_list_users():
    return jsonify({"users": ["admin", "user1", "user2"]})


@app.route("/admin/delete", methods=["DELETE"])    # VULN: no auth decorator
def admin_delete_user():
    user_id = request.args.get("id")
    return jsonify({"deleted": user_id})


# ── Debug Mode (CROSS-A05-001) ───────────────────────────────────────────────
if __name__ == "__main__":
    app.run(debug=True, host="0.0.0.0")   # VULN: debug=True in production code


# ── Logging Sensitive Data (CROSS-A09-001) ───────────────────────────────────
import logging
logger = logging.getLogger(__name__)

def authenticate(username, password):
    logger.info(f"Login attempt: user={username} password={password}")  # VULN: logs password
    return username == "admin"


# ── eval() usage (CROSS-A02-003) ─────────────────────────────────────────────
@app.route("/calculate")
def calculate():
    expression = request.args.get("expr")
    result = eval(expression)   # VULN: eval of user input
    return jsonify({"result": result})


# ── SSL Verification Disabled (CROSS-A05-002) ────────────────────────────────
import requests as req

def fetch_external(url):
    response = req.get(url, verify=False)   # VULN: no SSL verification
    return response.json()
